Rules/AvoidAlias.cs

```diff
@@ -104,38 +95,54 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
             IEnumerable<Ast> foundAsts = ast.FindAll(testAst => testAst is CommandAst, true);
 
             // Iterates all CommandAsts and check the command name.
-            foreach (Ast foundAst in foundAsts)
+            foreach (CommandAst cmdAst in foundAsts)
             {
-                CommandAst cmdAst = (CommandAst)foundAst;
-
                 // Check if the command ast should be ignored
                 if (IgnoreCommandast(cmdAst))
                 {
                     continue;
                 }
 
-                string aliasName = cmdAst.GetCommandName();
+                string commandName = cmdAst.GetCommandName();
 
                 // Handles the exception caused by commands like, {& $PLINK $args 2> $TempErrorFile}.
                 // You can also review the remark section in following document,
                 // MSDN: CommandAst.GetCommandName Method
-                if (aliasName == null
-                    || whiteList.Contains<string>(aliasName, StringComparer.OrdinalIgnoreCase))
+                if (commandName == null
+                    || whiteList.Contains<string>(commandName, StringComparer.OrdinalIgnoreCase))
                 {
                     continue;
                 }
 
-                string cmdletName = Helper.Instance.GetCmdletNameFromAlias(aliasName);
-                if (!String.IsNullOrEmpty(cmdletName))
+                string cmdletNameIfCommandNameWasAlias = Helper.Instance.GetCmdletNameFromAlias(commandName);
+                if (!String.IsNullOrEmpty(cmdletNameIfCommandNameWasAlias))
                 {
                     yield return new DiagnosticRecord(
-                        string.Format(CultureInfo.CurrentCulture, Strings.AvoidUsingCmdletAliasesError, aliasName, cmdletName),
+                        string.Format(CultureInfo.CurrentCulture, Strings.AvoidUsingCmdletAliasesError, commandName, cmdletNameIfCommandNameWasAlias),
                         GetCommandExtent(cmdAst),
                         GetName(),
                         DiagnosticSeverity.Warning,
                         fileName,
-                        aliasName,
-                        suggestedCorrections: GetCorrectionExtent(cmdAst, cmdletName));
+                        commandName,
+                        suggestedCorrections: GetCorrectionExtent(cmdAst, cmdletNameIfCommandNameWasAlias));
+                }
+
+                var isNativeCommand = Helper.Instance.GetCommandInfo(commandName) != null;

```

Comment:
This isn't necessarily a native command, but it's anything that will be returned by `Get-Command`. I was thinking that this was a bit open and wanted to be more specific about native apps, and there *is* an overload for GetCommandInfo which takes a CommandType, which would be perfect, because we could do this:
`var isNativeCommand = Helper.Instance.GetCommandInfo(commandName, (CommandTypes.Application|CommandTypes.ExternalScript))`
unfortunately, the internal method doesn't actually use this parameter if it's passed in which is a bug - **grrrr**

The reason I'm worried is that we may be still casting too broad a net here, and we'll get more false positives as the call as written will find everything named 'date' including scripts/functions/etc.

The long and short of it is there's no way to return *just* the native apps with the bug in the engine, so this may generate false positives. So the name of the variable is not what it purports to be (a native command)

Reply:
I will look into this.